src/win/fs.c

```diff
@@ -259,57 +310,112 @@ INLINE static void uv__fs_req_init(uv_loop_t* loop, uv_fs_t* req,
+                                       size_t w_source_len) {
+  WCHAR u;
+  WCHAR next;
 
-  if (target_len == 0) {
-    return -1;
+  u = w_source_ptr[0];
+  if (u >= 0xd800 && u <= 0xdbff && w_source_len > 1) {
+    next = w_source_ptr[1];
+    if (next >= 0xdc00 && next <= 0xdfff)
+      return 0x10000 + ((u - 0xd800) << 10) + (next - 0xdc00);
   }
+  return u;
+}
+
 
-  if (target_len_ptr != NULL) {
-    *target_len_ptr = target_len;
+static size_t fs__get_length_wide(const WCHAR* w_source_ptr,
+                                  size_t w_source_len) {
+  size_t target_len;
+  int32_t code_point;
+
+  target_len = 0;
+  for (; w_source_len; w_source_len--, w_source_ptr++) {
+    code_point = fs__get_surrogate_value(w_source_ptr, w_source_len);
+    /* Can be invalid UTF-8 but must be valid WTF-8. */
+    assert(code_point >= 0);
+    if (code_point < 0x80)
+      target_len += 1;
+    else if (code_point < 0x800)
+      target_len += 2;
+    else if (code_point < 0x10000)
+      target_len += 3;
+    else {
+      target_len += 4;
+      w_source_ptr++;
+      w_source_len--;
+    }
+  }
+  return target_len;
+}
+
+
+static int fs__wide_to_wtf8(WCHAR* w_source_ptr,
+                            size_t w_source_len,
+                            char** target_ptr,
+                            size_t* target_len_ptr) {
+  size_t target_len;
+  char* target;
+  int32_t code_point;
+
+  if (target_len_ptr == NULL || *target_len_ptr == 0) {
+    target_len = fs__get_length_wide(w_source_ptr, w_source_len);
+
+    if (target_len_ptr != NULL) {
+      *target_len_ptr = target_len;
+    }
+  } else {
+    target_len = *target_len_ptr;
   }
 
   if (target_ptr == NULL) {
     return 0;
   }
 
-  target = uv__malloc(target_len + 1);
-  if (target == NULL) {
-    SetLastError(ERROR_OUTOFMEMORY);
-    return -1;
+  if (*target_ptr == NULL) {
+    target = uv__malloc(target_len + 1);
+    if (target == NULL) {
+      SetLastError(ERROR_OUTOFMEMORY);
+      return -1;
+    }
+    *target_ptr = target;
+  } else {
+    target = *target_ptr;
+  }
+
+  for (; w_source_len; w_source_len--, w_source_ptr++) {
+    code_point = fs__get_surrogate_value(w_source_ptr, w_source_len);
+    /* Can be invalid UTF-8 but must be valid WTF-8. */
+    assert(code_point >= 0);
+
+    if (code_point < 0x80) {
+      *target++ = code_point;
+    } else if (code_point < 0x800) {
+      *target++ = 0xc0 | (code_point >> 6);
+      *target++ = 0x80 | (code_point & 0x3f);
+    } else if (code_point < 0x10000) {
+      *target++ = 0xe0 | (code_point >> 12);
+      *target++ = 0x80 | ((code_point >> 6) & 0x3f);
+      *target++ = 0x80 | (code_point & 0x3f);
+    } else {
+      *target++ = 0xf0 | (code_point >> 18);
+      *target++ = 0x80 | ((code_point >> 12) & 0x3f);
+      *target++ = 0x80 | ((code_point >> 6) & 0x3f);
+      *target++ = 0x80 | (code_point & 0x3f);
+      w_source_ptr++;
+      w_source_len--;
+    }
```

Comment:
What prevents a buffer overrun if `target` was a caller-provided buffer? Nothing, I think?

Reply:
Just the API not being used that way. I will change the API slightly to make it more explicit and add an assert.